neo/Ledger/Blockchain.cs

```diff
@@ -261,11 +272,22 @@ private RelayResultReason OnNewBlock(Block block)
                 return RelayResultReason.AlreadyExists;
             if (block.Index - 1 >= header_index.Count)
             {
-                block_cache_unverified[block.Index] = block;
+                AddUnverifiedBlockToCache(block);
                 return RelayResultReason.UnableToVerify;
             }
             if (block.Index == header_index.Count)
             {
+                if (block.Index > Height + 1)
```

Comment:
what if the block is 2 ahead of the current block though and it is missing transactions needed to make it verify?

Reply:
Won't that potentially make it fail verification erroneously and lose track of the block?

Reply:
The verification of transactions is done by the consensus nodes. Only blocks are verified here. And to verify the blocks, we only need the headers.

Reply:
but this code still executes even not on CN nodes; or even on a CN node the CN node could get 2 blocks behind right? .. then it could end up having the problem I'm describing.

Reply:
What I mean is that when the code is executed here, the verification of the transactions is complete, because you can verify the signature of the consensus nodes in the block header.

Reply:
but if `block.Index == header_index.Count` where this code is executing, then by definition you don't have the block header. So how could it be verifying it that way here?

Reply:
oh it is just looking at the previous block's hash and seeing if the block it got references the previous and is signed by CN nodes. Now I see what you are saying.

Reply:
So it shouldn't be possible to get something that would verify falsely here but in the future be valid. I guess this isn't needed then.

Reply:
One other questions, while we are talking about this; what happens if someone sends invalid header hashes? Where is that checked upon receiving headers? If it doesn't go in the `if (block.Index == header_index.Count` branch it just checks that that the hash matches, but I was having trouble finding the code that verifies the hashes are valid when it receives a message of header hashes. Can you point me to the code that would ensure that?

Reply:
Did not understand what you said...

Reply:
https://github.com/neo-project/neo/blob/2243d4b33eb74d5d31b1c7e64b6b9a16e9580bee/neo/Ledger/Blockchain.cs#L342-L365

Reply:
Thanks for linking that. It makes sense and looks robust.